### PR TITLE
Fix bugs in exp_count and add a boolean flag on whether to use both masks in calculating l_aux

### PR DESCRIPTION
**Pull request by yyyyyt123**
1. Correct the calculation of `exp_count` in top2gating.

2. Add a boolean flag on whether to use both masks in calculating `l_aux`

@awan-10 I'm so sorry that I made some misoperations in recorrecting the pr [#2551](https://github.com/microsoft/DeepSpeed/pull/2551). So I open a new pull request of same content as [#2551](https://github.com/microsoft/DeepSpeed/pull/2551).

### CODE REVIEW

File: deepspeed/moe/sharded_moe.py [hunk 1]
```diff
@@ -309,12 +310,18 @@ def top2gating(logits: Tensor,
     locations2 += torch.sum(mask1, dim=0, keepdim=True)
 
     # gating decisions
-    exp_counts = torch.sum(mask1, dim=0).detach().to('cpu')
+    exp_counts = torch.sum(mask1 + mask2, dim=0).detach().to('cpu')
```
Comment (ykim362) [inline deepspeed/moe/sharded_moe.py:313]: I think this needs to follow the config?

Comment (yyyyyt123) [reply deepspeed/moe/sharded_moe.py:313]: @ykim362  Thanks for your reply! But I think only the calculation of `l_aux` should follow the config. I thought the `exp_count` shows how many tokens are dispatched to the experts in different routing choices. I profiled the tokens dispatch pattern of the top1&top2 gate by printing the return value `self.deepspeed_moe.exp_counts` in `moe/layer.py/MoE/forward`. I found that in the `top2gating` function, the `exp_count` only counts the first routing choice (ignore the second routing choice). So I think `exp_count` might not have to follow the config?

Comment (yyyyyt123) [reply deepspeed/moe/sharded_moe.py:313]: I think in the `top2gating` function, the `exp_count` should count both the first&second expert choice of each token so that we can get the correct dispatch pattern. Feel free to correct me if I say something wrong.

**Comment by jomayeri** [comment]
@yyyyyt123 Please respond if you would like to keep this PR open. Otherwise I will close it. Thanks.
